provisioner/provision.js

```diff
@@ -0,0 +1,302 @@
+    // We could probably combine this with the .map of workerTypes below... meh...
+    debug('There are workerType definitions for these: %s', JSON.stringify(workerTypes.map(function(x) {
+      return x.workerType;
+    })));
+
+    return Promise.all(workerTypes.map(function(workerType) {
+      // We should be able to filter by a specific workerType
+      var wtDebug = 
+        _debug(baseDbgStr + ':workerType_' + workerType.workerType + ':run_' + that.__provRunId);
+      return that.provisionType(wtDebug, workerType, state, pricing);
+    }));
+  });
+
+  p = p.then(function(res) {
+    debug('Completed provisioning iteration');
+    return res;
+  });
+
+  return p;
+};
+
+
+
+/**
+ * Provision a specific workerType.  This promise will have a value of true if
+ * everything worked.  Another option is resolving to the name of the worker to
+ * make it easier to see which failed, but I'd prefer that to be tracked in the
+ * caller. Note that awsState as passed in should be specific to a workerType
+ */
+Provisioner.prototype.provisionType = function(debug, workerType, state, pricing) {
+  var that = this;
+
+  var p = this.Queue.pendingTasks(this.provisionerId, workerType.workerType);
+
+  p = p.then(function (pending) {
+    // Remember that we send the internally tracked state so that we can
+    // offset the count that we get here
+    var capacity = state.capacityForType(workerType, that.__outstanding);
+
+    if (typeof pending !== 'number') {
+      pending = 0;
+      debug('GRRR! Queue.pendingTasks(str, str) is returning garbage!  Assuming 0');
+    }
+
+    debug('capacity %d, pending: %d', capacity, pending);
+
+    if (capacity < workerType.maxCapacity) {
+      return workerType.provision(debug, pricing, capacity, pending);
+    } else {
+      // This is where we should kill excess capacity
+      return []
+    }
+
+  });
+
+  p = p.then(function(newSrs) {
+    that.addNewSpotRequests(debug, state, newSrs);
+    return newSrs;
+  });
+  
+  return p;
+};
+
+
+/**
+ * When we have too many instances or outstanding spot requests, we should kill
+ * spot requests and pending jobs.  We should also have a sanity threshold of
+ * maxCapacity * 2 which will start to kill running instances
+ */
+Provisioner.prototype.killExcess = function(debug, workerType, capacity) {
+  throw new Error('Implement me!');
+};
+
+/**
+ * Because the AWS is eventually consistent, it will sometimes take time for
+ * spot requests to show up in the describeSpotInstanceRequests calls for
+ * AWS state.  We will maintain an internal table of these submitted but
+ * not yet visible spot requests so that we can offset the count of a given
+ * instance type for figuring out running capacity.  If the provisioning
+ * process is restarted before the spot request shows up in the api's
+ * state we will lose track of it until it turns into an instance.
+ */
+Provisioner.prototype.addNewSpotRequests = function(debug, state, newSrs) {
```

Comment:
Again, keep state in `AwsState` and putting this into `AwsState` would mean that a lot of this code could be isolated. Rather than us maintaining and passing complex data structures around.